crates/p2p/src/swarm/handle.rs

```
@@ -45,3 +46,25 @@ where
         self.events.is_empty()
     }
 }
+
+impl<DSP> Stream for P2PHandle<DSP>
+where
+    DSP: prost::Message + Clone,
+{
+    type Item = Event<DSP>;
+
+    fn poll_next(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Self::Item>> {
+        let poll = Box::pin(self.next_event()).poll_unpin(cx);
+        match poll {
+            std::task::Poll::Ready(Ok(v)) => std::task::Poll::Ready(Some(v)),
+            std::task::Poll::Ready(Err(RecvError::Closed)) => std::task::Poll::Ready(None),
+            // NOTE FOR REVIEW: should we be silently swallowing skipped messages here or should we
+            // handle it differently?
```

Not sure if there is a way to handle the skipped messages from within the p2p crate itself. The best we can do is probably store it but that would mean the crate would be implicitly storing lagged messages. I think for now, it's fine to just log this and we can assess whether we need to handle this later.

---

Well we *can* propagate the error out of the handle. This implementation currently silently swallows the error. Do you prefer logging or propagating?

---

Either way, I don't know of a graceful way to handle skipped messages. That said, I'd prefer propagating the error out from here so that the caller can decide what to do with it with additional context.

---

Done.